src/server/gc_worker/gc_raftdb.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_rocks::raw::{new_compaction_filter_raw, CompactionFilter, CompactionFilterFactory};
+use engine_rocks::RAFT_LOG_GC_CONTEXT;
+use std::{collections::HashMap, ffi::CString};
+
+pub struct RaftLogCompactionFilterFactory {}
+
+impl RaftLogCompactionFilterFactory {}
+
+impl CompactionFilterFactory for RaftLogCompactionFilterFactory {
+    fn create_compaction_filter(
+        &self,
+        _: &engine_rocks::raw::CompactionFilterContext,
+    ) -> *mut engine_rocks::raw::DBCompactionFilter {
+        let ctx = RAFT_LOG_GC_CONTEXT.read().unwrap();
+        let filter = Box::new(RaftLogCompactionFilter::new(ctx.apply_idxs.clone()));
```

Comment:
This map could be pretty big, maybe instead of copying it, use a callback?

Reply:
Or just implement `CompactionFilter` for the context (`Arc<Context>`), I see the code logic here is quite simple.

Reply:
> This map could be pretty big, maybe instead of copying it, use a callback?

I'm afraid I did not get what 'callback' refers here. Do you mean use the ref of the map instead of copying it? I copy the map here to avoid the compaction blocking gc thread to update the map.

Reply:
By callback I means rlock the map during compaction, instead of copying the full volume every time. If blocking GC thread is an issue, you can move the map into filter factory, and keeping only the delta changes in context, which is merged back every time a new filter is created.

Reply:
OK, I will consider a better way to record the indexes.